Add governance patterns from GRC SIG (#288)
confidential-computing/governance: SIGs/GRC/publications/Verifier Governance.md

Reviewer: I think VHO needs to be redefined and replaced with VSO: Verifier System owner, where Verifier System comprises the hardware and operating system on which the Verifier Service runs.

Reviewer 2: For future PR.

# ![CCC GRC logo](./images/ccc_grc_logo.png) Verifier Governance

# Context

Confidential Computing requires a trusted Verifier **\[1\]**. This document covers the governance requirements inherent in operating trustworthy Verifiers. A properly governed Verifier must be subjected to a set of Confidential Computing-specific Control Objectives in order to satisfy the requirements of each Persona **\[3\]**.

# Problem

> A Verifier is the lynchpin of Confidential Computing – the Verifier is the ultimate arbiter of the identity of every Confidential Computing workload that relies on it, and any security vulnerability in the Verifier may thus lead to the entire edifice coming down. It is therefore extremely important to properly govern Verifier instances to ensure and assess their trustworthiness.

Reviewer 2: "the lynchpin of Confidential Computing" is a pretty strong statement.
Verifier failure doesn't break a TEE. 
It breaks remote reliance on interacting with what is thought to be a properly operating TEE. 
I wonder if this problem statement can be worded better to reflect these two aspects.

Reviewer: I put that statement in and I stand by it, but curious how would you reword it?
From the standpoint of the enterprise deploying a Confidential Computing ecosystem, how to secure and govern the Verifier is one of the most critical decisions, not unlike standing up an HSM or a Certificate Authority. Thus the wording in the document as it exists.

Reviewer 2: There's a passive use of confidential computing, where your workload has protection from other tenants or host software trying to inspect or change its memory. That's unaffected by a broken verifier.

The reliance on those protections being in place depends on the verifier and I totally agree with that. I also like the HSM and CA metaphors in your comment.

Maybe something like...

```suggestion
A Verifier plays a crucial role in Confidential Computing – it is the ultimate arbiter of the identity and security state of every Confidential Computing workload. A compromised Verifier may lead to the loss of assets entrusted to a Confidential Computing solution in the same manner that the compromise of a Hardware Security Module or Certificate authority would lead to insecure reliance and loss of assets. It is therefore extremely important to properly govern Verifier instances to ensure and assess their trustworthiness.
```

The discussion that follows will refer to the following Roles and Assets inherent in the Verifier space.

## Roles

| Role | Description and Trust Relationships |
| :---- | :---- |
| **Verifier Hardware Operator (VHO)** | The Cloud Service Provider (CSP) or a similar entity that owns and operates the hardware on which the Verifier Service runs. |
| **Verifier Service (VS)** | The entity that operates and bears responsibility for the secure operation of the Verifier Service on the hardware that VHO is responsible for. Trusts[^1] the VHO (but may choose to attest the Verifier hardware). |

> | **Verifier Tenant (VT)** | The entity that operates and bears responsibility for the portion of the Verifier Service hosting a given Verifier Tenant. Trusts the Verifier Service directly and VHO transitively, but may choose to attest Verifier Service and Verifier hardware as a precondition to such trust. |

Reviewer: I don't think this is a circular reference. Verifier Service is effectively the host where each Verifier Tenant has their own compartment. How would you reword it?

> | **Verifier Tenant (VT)** | The entity that operates and bears responsibility for the portion of the Verifier Service hosting a given Verifier Tenant. Trusts the Verifier Service directly and VHO transitively, but may choose to attest Verifier Service and Verifier hardware as a precondition to such trust. |

Reviewer 2: I'm commenting on the Verifier Tenant definition. It reads to me as "the entity that operates the service hosting a Tenant". I don't think you can say Tenant in the definition of Tenant. 
I think the VT is the relying party licensing a VS. The VT is responsible for defining the Appraisal Policies that the VS implements. 
At least I think that's the intent here

> | **Verifier Tenant (VT)** | The entity that operates and bears responsibility for the portion of the Verifier Service hosting a given Verifier Tenant. Trusts the Verifier Service directly and VHO transitively, but may choose to attest Verifier Service and Verifier hardware as a precondition to such trust. |

Reviewer: Pay attention to the word "portion". Yes, the portion of the overall Verifier service (perhaps running in a separate TEE but we don't specify that here) is what the VT is responsible for. The VS hosting the VT on the system managed by the VHO is just the control plane for our purposes. We might want to rely on it to keep the power on, or we may need to trust it if there is no TEE, but it is still an independent and isolated portion.

> | **Verifier Tenant (VT)** | The entity that operates and bears responsibility for the portion of the Verifier Service hosting a given Verifier Tenant. Trusts the Verifier Service directly and VHO transitively, but may choose to attest Verifier Service and Verifier hardware as a precondition to such trust. |

Reviewer: Let's discuss this in Wednesday's (May 7) GRC meeting.

![Solution diagram](./images/verifier_sol.png)

Verifier governance aims to achieve the following goals and ensure that the evidence is in place to prove it:

1. Trust in Verifier is established via the following mechanisms:  
   1. The Verifier Service MUST be properly administered: it runs expected code in a secure configuration on hardware properly administered by the VHO.  
   2. The Verifier Tenant MUST be executing the most current policies of that Tenant.  
   3. An Attester SHOULD **\[a\]** know which Verifier Tenant to contact and can authenticate it before sharing any data.  
        
2. High availability (i.e. in excess of the availability required of the applications which depend on the Verifier) SHOULD **\[b\]** be maintained, through well-understood mechanisms:  
   1. Maintaining multiple concurrent load-balanced Verifier instances  
   2. Replicating this load-balanced setup in separate geographic locations  
   3. Other common-sense measures, such as careful administration and constant monitoring for problems  
   4. Provisioning the Verifier with sufficient computing, networking and storage resources to avoid performance degradations and outages  

>    5. For multi-tenant Verifiers, imposing throttling, quotas, or other similar measures on individual Tenants to avoid noisy neighbor issues  

Reviewer 3: There's an additional aspect of verifier ability to provide services - being up to date with current sets of endorsements/reference values. This doc doesn't mention the relationship between verifiers and the Endorsers / Reference Value Provider supply chain. That also has a trust aspect in only admitting trusted RVs etc.

Reviewer: I agree that managing the trust relationships with suppliers of RVs should be mentioned. My preference would be to handle this as a separate Issue, with its own PR, however.

Reviewer 2: This seems like a pretty important aspect. We've had long discussions with customers about reference values and how timely policies need to be for enforcing the latest RVs. As a middle ground to deferring to another PR I'll take a crack at including at least some RV language and then if necessary we can improve upon it in a later draft. But if this PR is going to produce a public artifact it would be good to get RV/Endorsers mentioned even if incompletely.

Author: @TheBankster @SimonFrost-Arm can we resolve this if the changes proposed by @dcmiddle are acceptable?

Reviewer: Sure. There still should be an issue listed to version if necessary.

Reviewer 2: Will create issue

   2. Confidentiality offered to the histories or policies and cryptographic keys MUST match that offered to the active policies and cryptographic keys

# Governance Expectations Summary

The numbers in the left column below refer to **\[3\]**. Rows listed as N/A indicate that corresponding expectations are listed under different Patterns documents.

| \# | Description |
| :---- | :---- |
| 1-4, 6, 9-10, 13-15 | N/A |
| 5, 7-8 | Securely maintain a tamper-evident record of all Verifier policies. |
| 11 | Maintain high integrity and availability expected of a mission critical service. Maintain confidentiality of all secrets. Honor all current VS and VT policies. |
| 11-12 | Ensure secure administration of Verifier Hardware, Verifier Service and Verifier Tenant.Securely maintain and furnish on demand tamper-evident histories (per-tenant) of attestation decisions, within a specified SLA regarding retention period, timeliness, etc. |

# “SHOULD” vs. “MUST” Clarifications

> 1. This requirement is only present if the Tenant is concerned that Evidence submitted by the Attester during Remote Attestation is of a sensitive nature or if it directly uses the Attestation Results. Confidentiality of submitted Evidence and Attestation Results can be achieved via Transport Level Security between the Attester and the Verifier Tenant, with Verifier Tenant authenticated by the Attester prior to the Transport Level Security being established..  

Reviewer 2: nit typo ..
```suggestion
1. This requirement is only present if the Tenant is concerned that Evidence submitted by the Attester during Remote Attestation is of a sensitive nature or if it directly uses the Attestation Results. Confidentiality of submitted Evidence and Attestation Results can be achieved via Transport Level Security between the Attester and the Verifier Tenant, with Verifier Tenant authenticated by the Attester prior to the Transport Level Security being established.
```

# Problem

A Verifier plays a crucial role in Confidential Computing – it is the ultimate arbiter of the identity and security state of every Confidential Computing workload. A compromised Verifier may lead to the loss of assets intended for a Confidential Computing deployment in the same manner that the compromise of a Hardware Security Module or Certificate authority would lead to insecure reliance and loss of assets. It is therefore extremely important to properly govern Verifier instances to ensure and assess their trustworthiness.

The discussion that follows will refer to the following Roles and Assets inherent in the Verifier space.

## Roles

| Role | Description and Trust Relationships |
| :---- | :---- |
| **Verifier Hardware Operator (VHO)** | The Cloud Service Provider (CSP) or a similar entity that owns and operates the hardware on which the Verifier Service runs. |

> | **Verifier Service (VS)** | The entity that operates and bears responsibility for the secure operation of the Verifier Service on the hardware that VHO is responsible for. Trusts[^1] the VHO (but may choose to attest the Verifier hardware). |

Reviewer 2: Change wording so as not to define a term by using the term itself.
I'm somewhat uncertain if this should really be 2 definitions - one for the Verifier Service Operator and a separate definition for the Service. Since this section is called "roles" and to avoid snowballing changes, I'm sticking with just 1 definition that attempts to cover both the operator and its service.

```suggestion
| **Verifier Service (VS)** | The entity that operates and bears responsibility for the secure operation of a service to assess attestations. The VS may support a single customer or multiple customers defined here as Verifier Tenants. The VS trusts[^1] the VHO and may choose to minimize that trust by operating its Verifier Service Software within a Confidential Computing context such as a secure enclave process or confidential virtual machine. |
```

The list below assumes a multi-tenant Verifier (single-tenant is a trivial case so covered implicitly).

| Asset | Role | Description of the Asset | RC[^2] |
| :---- | :---: | :---- | :---: |
| **Verifier Service hardware** | VHO | Physical hardware on which the (multi-Tenant) Verifier Service runs | N/A |
| **Verifier Service software** | VS | Deployed executable code that constitutes the Verifier Service and is shared by all Verifier Tenants | No |
| **Verifier Service configuration** | VS | Configuration values that define the execution of the Verifier Service and integration with its supporting services (if any), *excluding* cryptographic keys | No |
| **Verifier Service secrets** | VS | Secrets (encryption, signing) utilized by the Verifier Service for its secure operation | Yes |

> | **Verifier Tenant authentication key** | VT | The private key (or similar) used to authenticate the Verifier Tenant to the Attester | Yes |

Reviewer 4: I wonder if this level of detail is necessary?  It sounds like we are prescribing a particular way of implementing authentication.  Perhaps it's sufficient to say all Verifier Tenant secrets (for authentication or encryption) need to be protected.

     
4. **System of Record Considerations:** Auditors and incident responders alike will care about the Verifier Service and Verifier Tenant histories, in addition to their present state. This includes the history and governance of policies (including both Verifier Service policies and policies related to the administration of the Verifier Tenant), and the history of Attestation Results.

# Solution

The terms MUST/SHOULD/MAY etc. below are used in accordance with **\[2\]**. Every SHOULD recommendation is explained separately in the “SHOULD vs. MUST Clarifications” section towards the end of this document.

![Solution diagram](./images/verifier_sol.png)

Verifier governance aims to achieve the following goals and ensure that the evidence is in place to prove it:

1. Trust in Verifier is established via the following mechanisms:  
   1. The Verifier Service MUST be properly administered: it runs expected code in a secure configuration on hardware properly administered by the VHO.  
   2. The Verifier Tenant MUST be executing the most current policies of that Tenant.  

>    3. An Attester SHOULD **\[a\]** know which Verifier Tenant to contact and can authenticate it before sharing any data.  

Reviewer 2: Per @SimonFrost-Arm 's feedback incorporate some RV/Endorser language.
```suggestion
   3. A Verifier Service MUST maintain current and accurate Reference Values and Endorser Identities (see [1] for definitions of these artifacts).
   4. An Attester SHOULD **\[a\]** know which Verifier Tenant to contact and can authenticate it before sharing any data.  
```

Reviewer 2: Create Issue for future PR

   1. Protect Verifier Tenant policies  
      1. Confidentiality SHOULD **\[e\]** be required  
      2. Integrity MUST be maintained  
   2. Appropriate lifecycle management of Verifier Tenant signing and encryption keys, as well as a tamper-proof auditable change log MUST be maintained  
      1. Verifier Tenants SHOULD **\[d\]** be allowed to bring their own keys for signing issued credentials and protecting sensitive Verifier Tenant policies and logs at-rest  
      2. Keys SHOULD **\[d\]** be rotated with some periodicity and MUST be rotated in case of suspected or actual compromise (**\[4\]**, **\[5\]**, **\[6\]**)  
   3. Confidentiality protections SHOULD **\[c\]** apply to Evidence and Attestation Results  
   4. Isolation from Verifier Service, especially in multi-tenant environments, SHOULD **\[e\]** be achieved  
   5. Peer tenant isolation MUST be achieved for a multi-tenant Verifier  
   6. Additional protections related to issues mentioned in the Privacy Considerations section of **\[1\]** MAY apply  
        
4. Ensure and be able to prove that the correct and current policies are applied at all times  

>    1. Tamper-evident logs of policy & cryptographic key histories, as well as associated attestation decisions, listing responsible actors (e.g., who has approved and/or applied each policy change, which policy was used for each decision), MUST be securely maintained  

Reviewer 2: Per @SimonFrost-Arm 's feedback incorporate some RV/Endorser language.
```suggestion
   1. Tamper-evident logs of Reference Values, Endorser Identities, VT Policies & cryptographic key histories, as well as associated attestation decisions, listing responsible actors (e.g., who has approved and/or applied each policy change, which policy was used for each decision), MUST be securely maintained  
```

Reviewer 2: Create issue
